lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+   */
+  static renderLine(dom, tmpl, line, lineOptions = {}) {
+    const {content, number, truncated} = line;
+
+    const clonedTemplate = dom.cloneTemplate('#tmpl-lh-code-snippet__line', tmpl);
+    const codeLine = dom.find('.lh-code-snippet__line', clonedTemplate);
+
+    if (lineOptions.highlight) {
+      codeLine.classList.add('lh-code-snippet__line--highlighted');
+    }
+    if (lineOptions.highlightMessage) {
+      codeLine.classList.add('lh-code-snippet__line--highlight-message');
+    }
+    if (lineOptions.code) {
+      codeLine.classList.add('lh-code-snippet__line--code');
+    }
+    if (lineOptions.collapsedOnly) {
+      codeLine.classList.add(SHOW_IF_COLLAPSED_CLASS);
+    }
+    if (lineOptions.expandedOnly) {
+      codeLine.classList.add(SHOW_IF_EXPANDED_CLASS);
+    }
+
+    const lineContent = content + (truncated ? '…' : '');
+    const lineContentEl = dom.find('.lh-code-snippet__line code', codeLine);
+    if (lineOptions.convertMarkdownLinkSnippets) {
+      lineContentEl.appendChild(dom.convertMarkdownLinkSnippets(lineContent));
+    } else {
+      lineContentEl.textContent = lineContent;
+    }
+
+    dom.find('.lh-code-snippet__line-number', codeLine).textContent = number.toString();
+
+    return codeLine;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippetHighlight} highlight
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
+      number: ' ',
+      content: highlight.message,
+    }, {
+      highlight: true,
+      highlightMessage: true,
+      convertMarkdownLinkSnippets: true,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineOptions} classOptions
+   * @return {Element}
+   */
+  static renderOmittedLines(dom, tmpl, classOptions = {}) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
+      number: '…',
+      content: '',
+    }, classOptions);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-code-snippet__content', tmpl);
+    const snippet = dom.find('.lh-code-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
+    nonLineSpecificHighlights.forEach(h =>
+      snippet.append(CodeSnippetRenderer.renderHighlightMessage(dom, tmpl, h))
+    );
+
+    snippet.append(CodeSnippetRenderer.renderSnippetLines(dom, tmpl, details));
+
+    // If expanded view still doesn't include all lines then indicate that that
+    const firstLineIsVisible = lines[0].number === 1;
+    const lastLineIsVisible = lines.slice(-1)[0].number === lineCount;
```

Comment:
clever! I'm not 100% convinced it's that much easier to grok than `lines[lines.length - 1].number` though ;)